<a href="https://colab.research.google.com/github/Karolya/AD_-/blob/main/HW_P.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

ДЗ 16.

Выполнить одно из заданий. Результат должен быть представлен в виде датафрейма. Все индексы (названия столбцов и строк) должны быть заданы явно. В работе обязательно использовать циклы. После  составления датафрейма применить к нему команду .loc для вывода на экран полученных данных (выбор данных – произвольный).
1. Составить код для заполнения таблицы умножения.
2. Составить код для заполнения таблицы сложения.
3. Составить код для заполнения таблицы степеней.
4. Составить код для заполнения таблицы простых чисел.

# Новый раздел

In [1]:
import pandas as pd
a=[]
for i in range(1,11):
    a.append(i**1)
ms=pd.Series(a)
ms.index=[1,2,3,4,5,6,7,8,9,10]

In [2]:
data=pd.DataFrame([ms**1,ms**2,ms**3,ms**4,ms**5,ms**6,ms**7,ms**8,ms**9,ms**10])
data

1     2      3        4        5         6          7           8   \
0   1     2      3        4        5         6          7           8   
1   1     4      9       16       25        36         49          64   
2   1     8     27       64      125       216        343         512   
3   1    16     81      256      625      1296       2401        4096   
4   1    32    243     1024     3125      7776      16807       32768   
5   1    64    729     4096    15625     46656     117649      262144   
6   1   128   2187    16384    78125    279936     823543     2097152   
7   1   256   6561    65536   390625   1679616    5764801    16777216   
8   1   512  19683   262144  1953125  10077696   40353607   134217728   
9   1  1024  59049  1048576  9765625  60466176  282475249  1073741824   

           9            10  
0           9           10  
1          81          100  
2         729         1000  
3        6561        10000  
4       59049       100000  
5      531441      1000000  
6     4782969     10000000  
7    43046721    100000000  
8   387420489   1000000000  
9  3486784401  10000000000

In [3]:
data[4].loc[3]

256

In [5]:
import pandas as pd
a=[]
for i in range(1,11):
  for n in range(1,11):
    a.append(i**n)
ms=pd.Series(a)
ms.index = [1,2,3,4,5,6,7,8,9,10]

ValueError: ignored

ДЗ 17


Домашнее заданиеЗарегистрироваться на сайте Kaggle: https://www.kaggle.com/.Скачать любой датасет. Изучить его. Составить описание используя команды .info() и .describe().


In [7]:
import pandas as pd
df1 = pd.read_csv("World University Rankings 2023.csv")
df1

University Rank                     Name of University        Location  \
0                  1                   University of Oxford  United Kingdom   
1                  2                     Harvard University   United States   
2                  3                University of Cambridge  United Kingdom   
3                  3                    Stanford University   United States   
4                  5  Massachusetts Institute of Technology   United States   
...              ...                                    ...             ...   
2336               -     University of the West of Scotland             NaN   
2337               -                  University of Windsor             NaN   
2338               -            University of Wolverhampton             NaN   
2339               -                University of Wuppertal             NaN   
2340               -    Xi’an Jiaotong-Liverpool University             NaN   

     No of student  No of student per staff International Student  \
0           20,965                     10.6                   42%   
1           21,887                      9.6                   25%   
2           20,185                     11.3                   39%   
3           16,164                      7.1                   24%   
4           11,415                      8.2                   33%   
...            ...                      ...                   ...   
2336           NaN                      NaN                   NaN   
2337           NaN                      NaN                   NaN   
2338           NaN                      NaN                   NaN   
2339           NaN                      NaN                   NaN   
2340           NaN                      NaN                   NaN   

     Female:Male Ratio OverAll Score  Teaching Score  Research Score  \
0              48 : 52          96.4            92.3            99.7   
1              50 : 50          95.2            94.8            99.0   
2              47 : 53          94.8            90.9            99.5   
3              46 : 54          94.8            94.2            96.7   
4              40 : 60          94.2            90.7            93.6   
...                ...           ...             ...             ...   
2336               NaN     34.0–39.2            24.1            15.5   
2337               NaN     34.0–39.2            35.1            29.4   
2338               NaN     34.0–39.2            18.2            14.3   
2339               NaN     34.0–39.2            26.4            26.7   
2340               NaN     34.0–39.2            17.8            14.8   

      Citations Score  Industry Income Score  International Outlook Score  
0                99.0                   74.9                         96.2  
1                99.3                   49.5                         80.5  
2                97.0                   54.2                         95.8  
3                99.8                   65.0                         79.8  
4                99.8                   90.9                         89.3  
...               ...                    ...                          ...  
2336             61.5                   37.9                         76.8  
2337             34.5                   44.2                         88.7  
2338             68.8                   37.3                         72.0  
2339             52.8                   52.1                         47.6  
2340             68.2                   38.2                         72.4  

[2341 rows x 13 columns]

In [8]:
df1.loc[df1["University Rank"]=="-"]

University Rank                   Name of University Location  \
2209               -                                  NaN      NaN   
2210               -                                  NaN      NaN   
2211               -                                  NaN      NaN   
2212               -                                  NaN      NaN   
2213               -                                  NaN      NaN   
...              ...                                  ...      ...   
2336               -   University of the West of Scotland      NaN   
2337               -                University of Windsor      NaN   
2338               -          University of Wolverhampton      NaN   
2339               -              University of Wuppertal      NaN   
2340               -  Xi’an Jiaotong-Liverpool University      NaN   

     No of student  No of student per staff International Student  \
2209           NaN                      NaN                   NaN   
2210           NaN                      NaN                   NaN   
2211           NaN                      NaN                   NaN   
2212           NaN                      NaN                   NaN   
2213           NaN                      NaN                   NaN   
...            ...                      ...                   ...   
2336           NaN                      NaN                   NaN   
2337           NaN                      NaN                   NaN   
2338           NaN                      NaN                   NaN   
2339           NaN                      NaN                   NaN   
2340           NaN                      NaN                   NaN   

     Female:Male Ratio OverAll Score  Teaching Score  Research Score  \
2209               NaN          56.9            81.0            72.9   
2210               NaN     51.2–54.3            55.7            46.8   
2211               NaN     47.0–48.7            31.1            23.7   
2212               NaN     42.1–44.9            32.5            45.3   
2213               NaN     42.1–44.9            45.3            41.7   
...                ...           ...             ...             ...   
2336               NaN     34.0–39.2            24.1            15.5   
2337               NaN     34.0–39.2            35.1            29.4   
2338               NaN     34.0–39.2            18.2            14.3   
2339               NaN     34.0–39.2            26.4            26.7   
2340               NaN     34.0–39.2            17.8            14.8   

      Citations Score  Industry Income Score  International Outlook Score  
2209              9.3                   99.1                         72.4  
2210             51.4                   99.1                         62.2  
2211             83.0                   75.5                         57.4  
2212             54.6                   61.6                         43.9  
2213             29.1                  100.0                         70.0  
...               ...                    ...                          ...  
2336             61.5                   37.9                         76.8  
2337             34.5                   44.2                         88.7  
2338             68.8                   37.3                         72.0  
2339             52.8                   52.1                         47.6  
2340             68.2                   38.2                         72.4  

[132 rows x 13 columns]

In [9]:
df1.loc[df1["Location"]=="United States"]

University Rank                       Name of University       Location  \
1                  2                       Harvard University  United States   
3                  3                      Stanford University  United States   
4                  5    Massachusetts Institute of Technology  United States   
5                  6       California Institute of Technology  United States   
6                  7                     Princeton University  United States   
...              ...                                      ...            ...   
1320       1201–1500  University of North Carolina Wilmington  United States   
1391       1201–1500                   Texas State University  United States   
1429       1201–1500              Western Michigan University  United States   
2038        Reporter        Northeast Ohio Medical University  United States   
2194        Reporter                 West Virginia University  United States   

     No of student  No of student per staff International Student  \
1           21,887                      9.6                   25%   
3           16,164                      7.1                   24%   
4           11,415                      8.2                   33%   
5            2,237                      6.2                   34%   
6            8,279                      8.0                   23%   
...            ...                      ...                   ...   
1320        15,940                     19.1                    2%   
1391        31,639                     19.3                    1%   
1429        17,612                     19.9                    8%   
2038           955                     11.8                    0%   
2194        25,500                     18.4                    5%   

     Female:Male Ratio OverAll Score  Teaching Score  Research Score  \
1              50 : 50          95.2            94.8            99.0   
3              46 : 54          94.8            94.2            96.7   
4              40 : 60          94.2            90.7            93.6   
5              37 : 63          94.1            90.9            97.0   
6              46 : 54          92.4            87.6            95.9   
...                ...           ...             ...             ...   
1320           64 : 36     18.4–24.3            18.5             8.1   
1391           59 : 41     18.4–24.3            19.1            16.2   
1429           52 : 48     18.4–24.3            22.7            15.2   
2038           54 : 46           NaN             NaN             NaN   
2194           52 : 48           NaN             NaN             NaN   

      Citations Score  Industry Income Score  International Outlook Score  
1                99.3                   49.5                         80.5  
3                99.8                   65.0                         79.8  
4                99.8                   90.9                         89.3  
5                97.3                   89.8                         83.6  
6                99.1                   66.0                         80.3  
...               ...                    ...                          ...  
1320             32.4                   37.3                         30.0  
1391             31.1                   39.8                         35.0  
1429             24.7                   37.3                         36.0  
2038              NaN                    NaN                          NaN  
2194              NaN                    NaN                          NaN  

[173 rows x 13 columns]

In [10]:
df1.describe()

No of student per staff  Teaching Score  Research Score  \
count              2208.000000     1799.000000     1799.000000   
mean                 19.000408       27.018010       23.016898   
std                  12.132224       13.282243       16.763819   
min                   0.400000       11.600000        7.400000   
25%                  12.600000       18.000000       11.300000   
50%                  16.600000       22.700000       17.000000   
75%                  22.200000       31.850000       28.900000   
max                 232.200000       94.800000       99.700000   

       Citations Score  Industry Income Score  International Outlook Score  
count      1799.000000            1799.000000                  1799.000000  
mean         48.495887              47.104558                    46.880378  
std          27.967185              15.093682                    22.582401  
min           0.800000              36.900000                    14.100000  
25%          23.100000              37.800000                    27.900000  
50%          47.200000              40.500000                    42.100000  
75%          72.350000              48.300000                    62.100000  
max         100.000000             100.000000                    99.700000

In [11]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2341 entries, 0 to 2340
Data columns (total 13 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   University Rank              2341 non-null   object 
 1   Name of University           2233 non-null   object 
 2   Location                     2047 non-null   object 
 3   No of student                2209 non-null   object 
 4   No of student per staff      2208 non-null   float64
 5   International Student        2209 non-null   object 
 6   Female:Male Ratio            2128 non-null   object 
 7   OverAll Score                1799 non-null   object 
 8   Teaching Score               1799 non-null   float64
 9   Research Score               1799 non-null   float64
 10  Citations Score              1799 non-null   float64
 11  Industry Income Score        1799 non-null   float64
 12  International Outlook Score  1799 non-null   float64
dtypes: float64(6), obj

In [12]:
df1.loc[df1["University Rank"]=="Reporter"]

University Rank                          Name of University  \
1697        Reporter  University Abdelhamid Ibn Badis Mostaganem   
1698        Reporter                     Abdullah Gül University   
1699        Reporter                          Abertay University   
1700        Reporter              Adamawa State University, Mubi   
1701        Reporter                     AECC University College   
...              ...                                         ...   
2204        Reporter                 Yogyakarta State University   
2205        Reporter                     York St John University   
2206        Reporter         Yusuf Maitama Sule University, Kano   
2207        Reporter       Zhytomyr Polytechnic State University   
2208        Reporter                         Ziauddin University   

            Location No of student  No of student per staff  \
1697         Algeria        24,493                     19.3   
1698          Turkey         2,509                     20.1   
1699  United Kingdom         4,010                     21.7   
1700         Nigeria        17,840                     70.0   
1701  United Kingdom           615                     15.4   
...              ...           ...                      ...   
2204       Indonesia        23,685                     18.5   
2205  United Kingdom         6,315                     18.6   
2206         Nigeria        12,880                     33.0   
2207         Ukraine         3,869                     15.4   
2208        Pakistan         4,906                      8.8   

     International Student Female:Male Ratio OverAll Score  Teaching Score  \
1697                    1%           67 : 33           NaN             NaN   
1698                    9%           43 : 57           NaN             NaN   
1699                   22%           48 : 52           NaN             NaN   
1700                    0%           35 : 65           NaN             NaN   
1701                   43%           47 : 53           NaN             NaN   
...                    ...               ...           ...             ...   
2204                    1%           59 : 41           NaN             NaN   
2205                   12%           65 : 35           NaN             NaN   
2206                    0%           48 : 52           NaN             NaN   
2207                    1%           34 : 66           NaN             NaN   
2208                    1%           63 : 37           NaN             NaN   

      Research Score  Citations Score  Industry Income Score  \
1697             NaN              NaN                    NaN   
1698             NaN              NaN                    NaN   
1699             NaN              NaN                    NaN   
1700             NaN              NaN                    NaN   
1701             NaN              NaN                    NaN   
...              ...              ...                    ...   
2204             NaN              NaN                    NaN   
2205             NaN              NaN                    NaN   
2206             NaN              NaN                    NaN   
2207             NaN              NaN                    NaN   
2208             NaN              NaN                    NaN   

      International Outlook Score  
1697                          NaN  
1698                          NaN  
1699                          NaN  
1700                          NaN  
1701                          NaN  
...                           ...  
2204                          NaN  
2205                          NaN  
2206                          NaN  
2207                          NaN  
2208                          NaN  

[512 rows x 13 columns]

В наборе данных представленно 2340 Университетов, для 293 Университета не указана локация.
Количество студентов указано для 94% Университетов.
132(5,6%) Универа не имеют рейтинг, 512(21,9%) имеют статус 'Reporter", итого  27,4% без рейтинга.
Только 76% Университетов имеют оценки (Общая оценка, Оценка преподавания, Оценка исследований, Оценка цитирования, Оценка дохода в отрасли,Оценка международных перспектив), я бы не стала рассматривать эти показатели для полноценного рейтинга.
Соотношение М:Ж (где известно) варьируется не сильно и этот показатель носит больше информационной характер.

Сильно выражена вариативность показателя "Количество персонала на учащихся", но поскольку нет уточнений преподовательский состав или АУП, то также выводы делать преждевременно.

Для полноценного анализа, необходимо более тщательное сравнение.





ДЗ 18
- проверить датасет на наличие пропусков, если они есть, вывести
строки с пропусками на экран;
- провести оценку корреляции между столбцами датафрейма и
сделать выводы;
- произвольно применить функции сложения, нахождения
среднего значения, медианы, минимума и максимума,
стандартного отклонения для случайной выборки из датасета;
- применить операции сравнения.

In [13]:
df=pd.DataFrame(df1)
df.isnull()==True

University Rank  Name of University  Location  No of student  \
0               False               False     False          False   
1               False               False     False          False   
2               False               False     False          False   
3               False               False     False          False   
4               False               False     False          False   
...               ...                 ...       ...            ...   
2336            False               False      True           True   
2337            False               False      True           True   
2338            False               False      True           True   
2339            False               False      True           True   
2340            False               False      True           True   

      No of student per staff  International Student  Female:Male Ratio  \
0                       False                  False              False   
1                       False                  False              False   
2                       False                  False              False   
3                       False                  False              False   
4                       False                  False              False   
...                       ...                    ...                ...   
2336                     True                   True               True   
2337                     True                   True               True   
2338                     True                   True               True   
2339                     True                   True               True   
2340                     True                   True               True   

      OverAll Score  Teaching Score  Research Score  Citations Score  \
0             False           False           False            False   
1             False           False           False            False   
2             False           False           False            False   
3             False           False           False            False   
4             False           False           False            False   
...             ...             ...             ...              ...   
2336          False           False           False            False   
2337          False           False           False            False   
2338          False           False           False            False   
2339          False           False           False            False   
2340          False           False           False            False   

      Industry Income Score  International Outlook Score  
0                     False                        False  
1                     False                        False  
2                     False                        False  
3                     False                        False  
4                     False                        False  
...                     ...                          ...  
2336                  False                        False  
2337                  False                        False  
2338                  False                        False  
2339                  False                        False  
2340                  False                        False  

[2341 rows x 13 columns]

In [14]:
df.isnull().sum()

University Rank                  0
Name of University             108
Location                       294
No of student                  132
No of student per staff        133
International Student          132
Female:Male Ratio              213
OverAll Score                  542
Teaching Score                 542
Research Score                 542
Citations Score                542
Industry Income Score          542
International Outlook Score    542
dtype: int64

In [15]:
df.corr()

<ipython-input-15-2f6f6606aa2c>:1: FutureWarning: The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  df.corr()


No of student per staff  Teaching Score  \
No of student per staff                     1.000000       -0.130569   
Teaching Score                             -0.130569        1.000000   
Research Score                              0.006686        0.882181   
Citations Score                            -0.005684        0.451169   
Industry Income Score                      -0.002588        0.492579   
International Outlook Score                 0.029470        0.385006   

                             Research Score  Citations Score  \
No of student per staff            0.006686        -0.005684   
Teaching Score                     0.882181         0.451169   
Research Score                     1.000000         0.514013   
Citations Score                    0.514013         1.000000   
Industry Income Score              0.583051         0.182949   
International Outlook Score        0.537993         0.536379   

                             Industry Income Score  \
No of student per staff                  -0.002588   
Teaching Score                            0.492579   
Research Score                            0.583051   
Citations Score                           0.182949   
Industry Income Score                     1.000000   
International Outlook Score               0.218060   

                             International Outlook Score  
No of student per staff                         0.029470  
Teaching Score                                  0.385006  
Research Score                                  0.537993  
Citations Score                                 0.536379  
Industry Income Score                           0.218060  
International Outlook Score                     1.000000

Какие страны и каким кол-ом универов представлены

In [16]:
df['Location'].value_counts()

United States     173
Japan             150
United Kingdom    149
India              91
China              82
                 ... 
Puerto Rico         1
Mozambique          1
Mauritius           1
Namibia             1
Kosovo              1
Name: Location, Length: 116, dtype: int64

In [17]:
df['Location'].mode()

0    United States
Name: Location, dtype: object

US лидер по кол-ву исследуемых ВУЗов

Минимальное и максимальное кол-во студентов на персонал

In [18]:
df['No of student per staff'].min(), df['No of student per staff'].max()

(0.4, 232.2)

In [19]:
df.sort_values('No of student per staff', ascending=False).head(1)

University Rank          Name of University      Location No of student  \
1121       1001–1200  University of South Africa  South Africa       236,282   

      No of student per staff International Student Female:Male Ratio  \
1121                    232.2                    3%           73 : 27   

     OverAll Score  Teaching Score  Research Score  Citations Score  \
1121     24.4–29.7            30.4            25.5             29.3   

      Industry Income Score  International Outlook Score  
1121                   37.6                         42.6

In [20]:
df.sort_values('No of student per staff', ascending=True).head(1)

University Rank                                 Name of University  \
1876        Reporter  The Graduate University for Advanced Studies, ...   

     Location No of student  No of student per staff International Student  \
1876    Japan           506                      0.4                   31%   

     Female:Male Ratio OverAll Score  Teaching Score  Research Score  \
1876           29 : 71           NaN             NaN             NaN   

      Citations Score  Industry Income Score  International Outlook Score  
1876              NaN                    NaN                          NaN

In [21]:
df['Location'].unique()


array(['United Kingdom', 'United States', 'Switzerland', nan, 'Canada',
       'Australia', 'Singapore', 'Japan', 'France', 'Sweden', 'China',
       'South Korea', 'Netherlands', 'Germany', 'Belgium', 'Finland',
       'Denmark', 'Austria', 'Norway', 'New Zealand', 'Spain', 'Italy',
       'Saudi Arabia', 'Luxembourg', 'Qatar', 'Brazil', 'Israel',
       'Ireland', 'Taiwan', 'India', 'United Arab Emirates',
       'Brunei Darussalam', 'Iceland', 'Lebanon', 'Philippines',
       'Portugal', 'Iran', 'Malaysia', 'Poland', 'Egypt', 'Turkey',
       'Greece', 'Vietnam', 'Algeria', 'Nigeria', 'Tanzania', 'Chile',
       'Pakistan', 'Ukraine', 'Romania', 'Czech Republic', 'South Africa',
       'Northern Cyprus', 'Hong Kong', 'Ethiopia', 'Jordan', 'Serbia',
       'Sri Lanka', 'Jamaica', 'Zambia', 'Iraq', 'Costa Rica', 'Cyprus',
       'Bangladesh', 'Mozambique', 'Colombia', 'Kenya', 'Namibia', 'Peru',
       'Latvia', 'Oman', 'Thailand', 'Lithuania', 'Slovenia', 'Uganda',
       'Malta', 'N

Сколько вариантов ВУЗов из Англии, США и Германии имеют рейтинг 3?
Для перевода столбца Рейтинг в числовые данные необходимо преобразовать тип дпнных, но для начала заменить все значения "-"

In [36]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2341 entries, 0 to 2340
Data columns (total 13 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   University Rank              2341 non-null   object 
 1   Name of University           2233 non-null   object 
 2   Location                     2047 non-null   object 
 3   No of student                2209 non-null   object 
 4   No of student per staff      2208 non-null   float64
 5   International Student        2209 non-null   object 
 6   Female:Male Ratio            2128 non-null   object 
 7   OverAll Score                1799 non-null   object 
 8   Teaching Score               1799 non-null   float64
 9   Research Score               1799 non-null   float64
 10  Citations Score              1799 non-null   float64
 11  Industry Income Score        1799 non-null   float64
 12  International Outlook Score  1799 non-null   float64
dtypes: float64(6), obj

In [46]:
df.loc[df['University Rank'] == '-', 'University Rank'] = '0'

In [64]:
df.loc[df['University Rank']].unique()

KeyError: ignored

А также необходимо изменить значения диапазонов рейтинга, для дальнейшего перевода в числовое значение

In [65]:
df

University Rank                     Name of University        Location  \
0                  1                   University of Oxford  United Kingdom   
1                  2                     Harvard University   United States   
2                  3                University of Cambridge  United Kingdom   
3                  3                    Stanford University   United States   
4                  5  Massachusetts Institute of Technology   United States   
...              ...                                    ...             ...   
2336               0     University of the West of Scotland             NaN   
2337               0                  University of Windsor             NaN   
2338               0            University of Wolverhampton             NaN   
2339               0                University of Wuppertal             NaN   
2340               0    Xi’an Jiaotong-Liverpool University             NaN   

     No of student  No of student per staff International Student  \
0           20,965                     10.6                   42%   
1           21,887                      9.6                   25%   
2           20,185                     11.3                   39%   
3           16,164                      7.1                   24%   
4           11,415                      8.2                   33%   
...            ...                      ...                   ...   
2336           NaN                      NaN                   NaN   
2337           NaN                      NaN                   NaN   
2338           NaN                      NaN                   NaN   
2339           NaN                      NaN                   NaN   
2340           NaN                      NaN                   NaN   

     Female:Male Ratio OverAll Score  Teaching Score  Research Score  \
0              48 : 52          96.4            92.3            99.7   
1              50 : 50          95.2            94.8            99.0   
2              47 : 53          94.8            90.9            99.5   
3              46 : 54          94.8            94.2            96.7   
4              40 : 60          94.2            90.7            93.6   
...                ...           ...             ...             ...   
2336               NaN     34.0–39.2            24.1            15.5   
2337               NaN     34.0–39.2            35.1            29.4   
2338               NaN     34.0–39.2            18.2            14.3   
2339               NaN     34.0–39.2            26.4            26.7   
2340               NaN     34.0–39.2            17.8            14.8   

      Citations Score  Industry Income Score  International Outlook Score  
0                99.0                   74.9                         96.2  
1                99.3                   49.5                         80.5  
2                97.0                   54.2                         95.8  
3                99.8                   65.0                         79.8  
4                99.8                   90.9                         89.3  
...               ...                    ...                          ...  
2336             61.5                   37.9                         76.8  
2337             34.5                   44.2                         88.7  
2338             68.8                   37.3                         72.0  
2339             52.8                   52.1                         47.6  
2340             68.2                   38.2                         72.4  

[2341 rows x 13 columns]

In [60]:
df['University Rank'] = df['University Rank'].astype(int)

ValueError: ignored

Поскольку нас интересует только ВУЗы с рейтингом "3", то достаточно будет узнать их количество

In [68]:
df[(df['Location'].isin(['United Kingdom', 'Germany', 'United States']))
    & (df['No of student per staff'] > 5)
    & (df['University Rank'] == '3')].shape[0]

2